### PR TITLE
High score list

### PR DESCRIPTION
**Pull request by pgaspar**
Prep for a new Array exercise

## TODO

- [x] Improve descriptions
- [x] Do these tests make sense? Do we need so many?
- [x] Add content to description.md
- [x] Add content to metadata.yml (the source_url field is not required)
- [x] canonical-data.json seems to be invalid (run `yarn test`)
- [x] Come up with a better name for `top_3` specs
- [x] Re-organize specs
- [x] Add spec for `scores`
- [x] Squash commits
- [x] Have another look at the tests - are some of them redundant?

### CODE REVIEW

File: exercises/high-scores/canonical-data.json [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        },
+        {
+          "description": "Returns personal bests from a long list",
+          "property": "top_3",
+          "input": {
+            "scores": [1, 3, 9, 3, 10, 2, 1, 0, 3, 4, 4, 7, 7]
+          },
+          "expected": [10, 9, 7]
+        }
+      ]
```
Comment (emcoding) [inline exercises/high-scores/canonical-data.json:72]: It this stray bracket causing trouble?

**Comment by emcoding** [comment]
> Do these tests make sense? Do we need so many?

I realize now that ideally, the tests would help the student build up the solution. In real TDD mode: they're writing the code that solves the next test. That means: the order of the tests matter. And maybe that will reveal that we'll need even one or two more. Or that some can go.

File: exercises/high-scores/canonical-data.json (hunk 1 above)
Comment (pgaspar) [reply exercises/high-scores/canonical-data.json:72]: Turns out it's because of `"property": "top_3",` 😛. The Readme says:

> The 'property' is a string in lowerCamelCase identifying the type of test, but most of the times it is just the name of a function being tested.

So, we're going to have a bit of trouble with the `top_3` method after all. `top3` fails too. `topThree` works. What do you think? `top` only? Something else?

Comment (emcoding) [reply exercises/high-scores/canonical-data.json:72]: `topTriad` or `Triad` :-)  

triad: "a group or set of three related people or things."

**Comment by emcoding** [comment]
After you managed to sort out the canonical_data file, shall I do the rearrangement of the tests?

File: exercises/high-scores/canonical-data.json (hunk 1 above)
Comment (pgaspar) [reply exercises/high-scores/canonical-data.json:72]: But also, the first thing that comes up on google under Triad:

> A triad is one of many branches of Chinese transnational organized crime syndicates based in China, Hong Kong...

Other ideas:

* `topScores`
* `topTrio` (don't like this very much either - might as well just use `topThree`)
* `leadScores`
* `best`

If you don't like any of these we can go with whatever you prefer and see if someone has better ideas when we submit the PR.

I'm going with `top` right now just so it validates and works directly with our test generator.

**Comment by pgaspar** [comment]
> After you managed to sort out the canonical_data file, shall I do the rearrangement of the tests?

What order were you thinking? Feel free to sort them on the canonical_data file whenever you want :+1:

File: exercises/high-scores/canonical-data.json (hunk 1 above)
Comment (emcoding) [reply exercises/high-scores/canonical-data.json:72]: Thanks for catching that. I'm okay with `top`.

**Comment by emcoding** [comment]
- I changed the order of a few tests. Can you check please? 
- I'd vote for removing the test for personal bests with a zero, to avoid discussion/confusion on edge cases, and we don't handle zero's other than non-zero scores.
- I'm okay with removing the tests for personal bests less than 3.  I can't think of a way that students try to ask for 3 elements that will fail. OTOH, it's a good illustration of what Ruby takes care of for us. So let's leave it in (and I left this comment for future reference.)

If you're okay with the new order, I'll change them in the Ruby file too.

Almost there!

**Comment by pgaspar** [comment]
Everything sounds and looks good.

To update the ruby test file run the generator script, as it also updates the commit hash:

```
# On the ruby repo:
bin/generate high-scores
```

I can also run it for you if you are editing everything directly on github.

**Comment by emcoding** [comment]
Great if you can run it for me!!  Then I can try to finish it off now, and then we're done here, I think?

**Comment by pgaspar** [comment]
@F3PiX I think we're ready to open the PR on problem specifications. I've added you as a co-author on the commit 🎉

Do you want to open this PR or should I?

**Comment by pgaspar** [comment]
Oh, we didn't address this:

> * I'd vote for removing the test for personal bests with a zero, to avoid discussion/confusion on edge cases, and we don't handle zero's other than non-zero scores.

I thought you had removed it when you switched the order. We can still remove it if you want.

File: exercises/high-scores/canonical-data.json [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+          "expected": [5, 3, 1]
+        },
+        {
+          "description": "Returns personal bests including zero",
+          "property": "top",
+          "input": {
+            "scores": [2, 0, 3]
+          },
+          "expected": [3, 2, 0]
+        },
```
Comment (pgaspar) [inline exercises/high-scores/canonical-data.json:47]: The goal here wasn't exactly the 0. It was the first (in the TDD sense) where the order was not the same as the initial order.

But maybe we just remove "Returns personal bests when last is less"? We can also change 0 to be something else.

File: exercises/high-scores/canonical-data.json [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+      "expected": "Your latest score was 7. That's your personal best!"
+    },
+    {
+      "description": "Returns the goal when highest is latest",
+      "property": "goal",
+      "input": {
+        "scores": [2, 4, 0, 3, 7]
+      },
+      "expected": "Your latest score was 7. That's your personal best!"
+    },
```
Comment (pgaspar) [inline exercises/high-scores/canonical-data.json:81]: We can probably remove this spec, don't you think? (or the one before)

File: exercises/high-scores/canonical-data.json [hunk 4]
```diff
@@ -62,7 +62,7 @@
           "expected": [4]
         },
         {
-          "description": "Returns personal bests from a long list",
+          "description": "Returns three top scores from a long list",
```
Comment (pgaspar) [inline exercises/high-scores/canonical-data.json:65]: Should we also change the other tests that have "personal bests" in their descriptions? (Also maybe change phrasing to "top three scores"?)

Comment (emcoding) [reply exercises/high-scores/canonical-data.json:65]: I didn't change those, because, in TDD manner, they can just return the `scores.sort.reverse` there. It's not until here that they have to `take(3)`.

Comment (pgaspar) [reply exercises/high-scores/canonical-data.json:65]: Ah, interesting. Ok 👍
